### PR TITLE
Coverage for Lua (Windows)

### PR DESCRIPTION
**Pull request by blueyed**

### CODE REVIEW

**Comment by blueyed** [comment]
New test failure (?), maybe due to coverage (slower):
```
[  ERROR   ] ...s/build/neovim/neovim/test/functional/ui/output_spec.lua @ 53: shell command :! throttles shell-command output greater than ~10KB
test/functional/ui/screen.lua:554: Failed to match any screen lines.
Expected (anywhere): "
%."
Actual:
  |XXXXXXXXXX 572                                    |
  |XXXXXXXXXX 573                                    |
  |XXXXXXXXXX 574                                    |
  |XXXXXXXXXX 575                                    |
  |                                                  |
  |                                                  |
  |{3:-- TERMINAL --}                                    |
```

**Comment by blueyed** [comment]
Lua files included there now: https://codecov.io/gh/neovim/neovim/compare/84c1d3aedeb882f07bd7ff029d4ce91c50bdb9eb...daa2034eda1091bd030a2e6058e20ea41d8fba23/tree

**Comment by blueyed** [comment]
If this makes sense to you, I'd also add it for AppVeyor.
I've thought about configuring this with CMake, but `-DUSE_LUACOV` would only be passed through to RunTests.cmake then.  I think it is OK for now to require manual setup of this (install luacov/cluacov (the latter is for performance), and pass `--coverage` (`BUSTED_ARGS`)).

**Comment by blueyed** [comment]
I think having it in a (faster) non-gcov job might make sense (to spread the slowness due to coverage tracking).

**Comment by justinmk** [comment]
This is a very low priority so I'd rather not spend time on this.

**Comment by blueyed** [comment]
ok, so not for Windows additionally?
I think it is quite ready already, and took not much effort.

**Comment by bfredl** [comment]
What lua files is this meant to cover? It seems to mostly test the `_spec.lua` files (rather than built-in lua or runtime) which are not very useful.

**Comment by blueyed** [comment]
`runtime/lua` is partially covered (inspect is missing there).
`src/nvim/lua/vim.lua` for example cannot be covered (easily) I guess.
It's just a best effort approach, and I believe that having tests included with coverage is good in general, because it helps you being aware of not executed code/tests therein.

I've not looked into it in detail, but just consider it a small helper when it comes to diff coverage with PRs for example.

**Comment by blueyed** [comment]
btw: I've done this to get (some/better) feedback with/on https://github.com/neovim/neovim/pull/10421 etc.

**Comment by blueyed** [comment]
It would help e.g. with https://github.com/neovim/neovim/pull/10773 and related things to ensure coverage (of tests) does not drop.

File: .luacov [hunk 1]
```diff
@@ -0,0 +1,18 @@
+-- Configuration file for LuaCov
+
+local source = os.getenv('PWD')
+
+local function escape_pattern(str)
+    return string.gsub(str, "%W", "%%%1")
+end
```
[the remark this answers is not in the document]
Comment (blueyed) [reply .luacov]: Will copy it - not sure it should / can be `required()` easily.

File: ci/before_script.sh [hunk 2]
```diff
@@ -35,5 +35,10 @@ fi
 # Compile dependencies.
 build_deps
 
+# Install cluacov for Lua coverage (-DUSE_LUACOV).
+if [[ "$DO_LUA_COVERAGE" == 1 ]]; then
```
[the remark this answers is not in the document]
Comment (blueyed) [reply ci/before_script.sh]: > E.g. why is it GCOV in one case, but DO_FOO in another?

￼`$GCOV` is referring to the "gcov" program that should be used, and used by other tools by convention also (e.g. `gcovr` IIRC).
So DO_GCOV means that C coverage should be done, while GCOV refers to the gcov program.  But of course "not empty $GCOV" can be used to replace "DO_GCOV".

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
Surprising that aggregate coverage increased. Note that https://github.com/mpeterv/cluacov is likely to be unmaintained since the author died recently :(

**Comment by blueyed** [comment]
Just for reference - contacted codecov support about it:

It seems like Lua coverage for Windows builds is not merged properly.

It is missing in the report at https://codecov.io/gh/neovim/neovim/src/7435ae4b87cb78d53e7177fee0731e32f613917b/test/functional/helpers.lua#L665...668.

Build: https://codecov.io/gh/neovim/neovim/commit/7435ae4b87cb78d53e7177fee0731e32f613917b/build
Job: 352a89cov3ncxk29
Uploaded data: https://codecov.io/codecov/v4/raw/2019-09-30/A8E2A0366253881A56434672B408D166/7435ae4b87cb78d53e7177fee0731e32f613917b/88cc8884-43d4-4328-adeb-a1c43deb29b5.txt


          -- Helper to skip tests. Returns true in Windows systems.
          -- pending_fn is pending() from busted
        1 function module.pending_win32(pending_fn)
       60   if iswin() then
       30     if pending_fn ~= nil then
       30       pending_fn('FIXME: Windows', function() end)
              end
       11     return true
            else
********0     return false
            end
          end

The file is named "test\functional\helpers.lu" in the upload.

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
AppVeyor ran twice as long, hope that isn't going to always happen...

**Comment by blueyed** [comment]
> AppVeyor ran twice as long, hope that isn't going to always happen...

Was not using cluacov there..  amended.

**Review by justinmk — APPROVED** [review]

**Comment by blueyed** [comment]
Hmm.. took still 29 min 36 sec..

**Comment by blueyed** [comment]
Now it took only 19 min 11 sec, which is normal.

**Pull request merged**